### PR TITLE
gnrc_sixlowpan_iphc: add fragment forwarding stubs

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
... when VRB is available

This is once again one of these tiny PRs that belong to a bigger PR, @benpicco is always talking about. However, in this case it was already introduced in an earlier version to #11068 (edc9b42051c23f058333673e192afa89b41a02bd) but is now needed also for SFR. So instead of having this piece of code copy-pasted in both, I decided to split it out (even though it is not functional without it).

### Testing procedure
There is a test application `tests/gnrc_sixlowpan_iphc_w_vrb` in this PR that should work on any board that it fits on:

```
make -C tests/gnrc_sixlowpan_iphc_w_vrb flash test
```

Confirm that a fragmented and unfragmented UDP packet can still be sent between two 6LoWPAN nodes.

Additionally, you can check-out #11068 and

1. confirm in its history that that PR only adds the actual minimal forwarding steps to `gnrc_sixlowpan_iphc.c`
2. Run its tests `make -C tests/gnrc_sixlowpan_frag_minfwd flash test` on a board they fit on

### Issues/PRs references
Taken out of #11068's edc9b42 (as it is also required for other 6LoWPAN fragment forwarding schemes like SFR). It is however completely independent of that.

Route to 6LoWPAN fragment forwarding

### CODE REVIEW

**Comment by miri64** [comment]
I also added a subroutine to do the actual forwarding in now. This code will be common between #11068 and SFR, so it makes sense to have a common base PR. Currently the function doesn't do anything except for returning `-ENOTSUP`, but I will provide some patches so it can be tested.

**Comment by miri64** [comment]
I started a test application. Up until now, only the success case is tested, I plan however to add some failure cases. When setting `ENABLE_DEBUG=1` in `sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c` one should see something like the following for now:

```
main(): This is RIOT! (Version: 2020.01-devel-461-g6d46e-gnrc_sixlowpan_iphc/enh/vrb-creation)
.6lo iphc: found route, trying to forward
6lo iphc: Do not know how to forward fragment from (2A:AB:DC:15:54:01:64:79, 15) to (4A:3D:1D:0C:98:31:58:AE, 1)
Original fragmentation header:
00000000  C0  BC  00  0F
IPHC headers + payload:
00000000  7A  00  3A  20  01  0D  B8  00  00  00  00  00  00  00  00  00
00000010  00  00  01  20  01  0D  B8  00  00  00  00  00  00  00  00  00
00000020  00  00  02
00000000  80  00  8E  A0  23  8F  00  02  9D  4B  B2  1C  53  53  53  53
00000010  53  53  53  53  53  53  53  53  53  53  53  53  53  53  53  53
00000020  53  53  53  53  53  53  53  53  53  53  53  53  53  53  53  53
00000030  53  53  53  53  53  53  53  53
packet buffer: first byte: 0x565dc120, last byte: 0x565dc920 (size: 2048)
  position of last byte used: 512
~ unused: 0x565dc120 (next: (nil), size: 2048) ~

OK (1 tests)
```

**Comment by miri64** [comment]
Finished tests and squashd. No longer WIP.

**Comment by miri64** [comment]
Updated the testing procedures

**Comment by miri64** [comment]
Not really sure why Murdock is failing on `native` here... I was not able to reproduce this failure locally, neither with GCC nor mit LLVM and I tried it on two systems.

<details><summary>Ubuntu 19.10</summary><pre>
  Operating System Environment
-----------------------------
       Operating System: "Arch Linux" 
                 Kernel: Linux 5.3.10-arch1-1 x86_64 unknown

Installed compiler toolchains
-----------------------------
             native gcc: gcc (GCC) 9.2.0
      arm-none-eabi-gcc: arm-none-eabi-gcc (Arch Repository) 9.2.0
                avr-gcc: avr-gcc (GCC) 9.2.0
       mips-mti-elf-gcc: missing
             msp430-gcc: missing
   riscv-none-embed-gcc: missing
   xtensa-esp32-elf-gcc: xtensa-esp32-elf-gcc (crosstool-NG crosstool-ng-1.22.0-80-g6c4433a) 5.2.0
   xtensa-lx106-elf-gcc: missing
                  clang: clang version 9.0.0 (tags/RELEASE_900/final)

Installed compiler libs
-----------------------
   arm-none-eabi-newlib: "3.1.0"
    mips-mti-elf-newlib: missing
riscv-none-embed-newlib: missing
xtensa-esp32-elf-newlib: "2.2.0"
xtensa-lx106-elf-newlib: missing
               avr-libc: "2.0.0" ("20150208")

Installed development tools
---------------------------
                  cmake: cmake version 3.15.5
               cppcheck: missing
                doxygen: 1.8.16
                    git: git version 2.24.0
                   make: GNU Make 4.2.1
                openocd: Open On-Chip Debugger 0.10.0+dev-00746-g8f518d35 (2019-03-21-09:43)
                 python: Python 3.7.4
                python2: Python 2.7.17
                python3: Python 3.7.4
                 flake8: 3.6.0 (mccabe: 0.6.1, pycodestyle: 2.4.0, pyflakes: 2.0.0) CPython 3.7.4 on Linux
             coccinelle: missing
  </pre>
  </details>
<details><summary>An Arch Linux</summary><pre>
<details><summary>An Arch Linux</summary><pre>
  Operating System Environment
-----------------------------
       Operating System: "Arch Linux" 
                 Kernel: Linux 5.3.10-arch1-1 x86_64 unknown

Installed compiler toolchains
-----------------------------
             native gcc: gcc (GCC) 9.2.0
      arm-none-eabi-gcc: arm-none-eabi-gcc (Arch Repository) 9.2.0
                avr-gcc: avr-gcc (GCC) 9.2.0
       mips-mti-elf-gcc: missing
             msp430-gcc: missing
   riscv-none-embed-gcc: missing
   xtensa-esp32-elf-gcc: xtensa-esp32-elf-gcc (crosstool-NG crosstool-ng-1.22.0-80-g6c4433a) 5.2.0
   xtensa-lx106-elf-gcc: missing
                  clang: clang version 9.0.0 (tags/RELEASE_900/final)

Installed compiler libs
-----------------------
   arm-none-eabi-newlib: "3.1.0"
    mips-mti-elf-newlib: missing
riscv-none-embed-newlib: missing
xtensa-esp32-elf-newlib: "2.2.0"
xtensa-lx106-elf-newlib: missing
               avr-libc: "2.0.0" ("20150208")

Installed development tools
---------------------------
                  cmake: cmake version 3.15.5
               cppcheck: missing
                doxygen: 1.8.16
                    git: git version 2.24.0
                   make: GNU Make 4.2.1
                openocd: Open On-Chip Debugger 0.10.0+dev-00746-g8f518d35 (2019-03-21-09:43)
                 python: Python 3.7.4
                python2: Python 2.7.17
                python3: Python 3.7.4
                 flake8: 3.6.0 (mccabe: 0.6.1, pycodestyle: 2.4.0, pyflakes: 2.0.0) CPython 3.7.4 on Linux
             coccinelle: missing
  </pre>
  </details>
  </pre>
  </details>

**Comment by miri64** [comment]
Even when build in docker they pass.....

**Comment by miri64** [comment]
Ah! When I rebase (not *merge*) to current master I get the same error Oo. I try to find out, what's happening.

**Comment by miri64** [comment]
Ahhhhh, due to #12698 fixing the remove function of the VRB, the VRB entry is actually removed, as `_forward_frag()` in `gnrc_sixlowpan_iphc.c` still returns `-ENOTSUP`, causing the VRB entry to be removed again. This is of course a broken test design, so I fixed-up the tests.

**Comment by miri64** [comment]
Ping @cgundogan?

**Comment by benpicco** [comment]
`tests/gnrc_sixlowpan_iphc_w_vrb` is [still](https://ci.riot-os.org/RIOT-OS/RIOT/12629/b0aa053d7d7003eebe55af15efc414fc2238fe98/output/compile/tests/gnrc_sixlowpan_iphc_w_vrb/native:llvm.txt) [failing](https://ci.riot-os.org/RIOT-OS/RIOT/12629/b0aa053d7d7003eebe55af15efc414fc2238fe98/output/compile/tests/gnrc_sixlowpan_iphc_w_vrb/native:gnu.txt).

Also no review on this one yet, so I guess it's easier for everyone if you just squash.

**Comment by miri64** [comment]
Mhhhhh... they *do* work on my machine :-/

**Comment by miri64** [comment]
Ah, there might be something going wrong with current master rebased. Will check!

**Comment by miri64** [comment]
First of all: squashed!

**Comment by miri64** [comment]
(and rebased, though this wasn't my intention ^^")

**Comment by miri64** [comment]
Fixed. The reason was that the arrival time of the VRB entries that filled the VRB was not set for the VRB-full test, so the garbage collection just through them out when the new entry was created.

**Comment by miri64** [comment]
Not sure why this worked with the old version of this PR... But now its fixed so :woman_shrugging:

**Comment by miri64** [comment]
Another update to `Makefile.ci` (hopefully the last) and squashed directly. Would be great to have this in soonish. Though I tried to make it not to conflict too much with #12879, merging the FF branches with that one was quite hellish ^^"

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c [hunk 1]
```diff
@@ -92,6 +96,12 @@
 #define NHC_UDP_8BIT_PORT           (0xF000)
 #define NHC_UDP_8BIT_MASK           (0xFF00)
 
+/* currently only used with forwarding output, remove guard if more debug info
+ * is added */
+#ifdef MODULE_GNRC_SIXLOWPAN_FRAG_VRB
```
Comment (benpicco) [inline sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c]:
```suggestion
#if defined(MODULE_GNRC_SIXLOWPAN_FRAG_VRB) && ENABLE_DEBUG
```

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c [hunk 2]
```diff
@@ -626,11 +640,25 @@ void gnrc_sixlowpan_iphc_recv(gnrc_pktsnip_t *sixlo, void *rbuf_ptr,
         rbuf->super.current_size += (uncomp_hdr_len - payload_offset);
 #ifdef MODULE_GNRC_SIXLOWPAN_FRAG_VRB
         if (vrbe != NULL) {
+            int res = -1;
             DEBUG("6lo iphc: found route, trying to forward\n");
             ipv6_hdr->hl--;
             vrbe->super.current_size = rbuf->super.current_size;
-            /* TODO: forward here according to forwarding scheme here */
```
Comment (benpicco) [inline sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c]: Just leave the comment but drop the TODO - makes it easier to understand what's going on.

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c [hunk 3]
```diff
@@ -644,6 +672,74 @@ void gnrc_sixlowpan_iphc_recv(gnrc_pktsnip_t *sixlo, void *rbuf_ptr,
...
+        return res;
+    }
+    else {
+        DEBUG("6lo iphc: unable to compress packet for forwarding\n");
+        gnrc_pktbuf_release(decoded_pkt);
+        return NULL;
+    }
+}
+
+static int _forward_frag(gnrc_pktsnip_t *pkt, gnrc_pktsnip_t *frag_hdr,
```
Comment (benpicco) [inline sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c]: Please add a comment that this is just a dummy implementation.

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c (hunk 1 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c]: Nope, that will lead to compilation errors, as `addr_str` is also "used" with `ENABLE_DEBUG=0` (`DEBUG` is always compiled in, but thrown out by the optimizer, as is this string array, as it is not used in this case). This is the case for quite a while now. See https://github.com/RIOT-OS/RIOT/pull/5166.

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c (hunk 3 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c]: It's not even an implementation. It just provides debug output ATM (both #11068 and #12303 amend to this function though). I will however reflect that in a comment.

**Comment by miri64** [comment]
Addressed comments and squashed directly as otherwise I wouldn't have been able to adapt the comment that was removed in a later commit.

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c (hunk 2 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c]: Done

File: sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c (hunk 3 above)
Comment (miri64) [reply sys/net/gnrc/network_layer/sixlowpan/iphc/gnrc_sixlowpan_iphc.c]: Done

**Comment by miri64** [comment]
Thanks for the review!

**Pull request merged**